crypto/fips/self_test.c

```diff
@@ -0,0 +1,349 @@
+{
+    if (ev != NULL && SELF_TEST_cb != NULL) {
+        ev->phase = SELF_TEST_PHASE_CORRUPT;
+        if (!SELF_TEST_cb(ev->params))
+            bytes[0] ^= 1;
+    }
+}
+
+/*
+ * For RSA key generation it is not known whether the key pair will be used
+ * for key transport or signatures. FIPS 140-2 IG 9.9 states that in this case
+ * either a signature verification OR an encryption operation may be used to
+ * perform the pairwise consistency check. The simpler encrypt/decrypt operation
+ * has been chosen for this case.
+ */
+int SELF_TEST_keygen_pairwise_test_rsa(RSA *rsa)
+{
+    int ret = 1;
+    unsigned int ciphertxt_len;
+    unsigned char *ciphertxt;
+    const unsigned char plaintxt[16] = {0};
+    unsigned char decoded[256];
+    unsigned int decoded_len;
+    unsigned int plaintxt_len = (unsigned int)sizeof(plaintxt_len);
+    int padding = RSA_PKCS1_PADDING;
+    ST_EVENT ev;
+
+    SELF_TEST_EVENT_init(&ev);
+    SELF_TEST_EVENT_onbegin(&ev, SELF_TEST_TYPE_PCT,
+                            SELF_TEST_DESC_PCT_RSA_PKCS1);
+
+    ciphertxt_len = RSA_size(rsa);
+    ciphertxt = OPENSSL_zalloc(ciphertxt_len);
+    if (ciphertxt == NULL) {
+        goto err;
+    }
+
+    ciphertxt_len = RSA_public_encrypt(plaintxt_len, plaintxt, ciphertxt, rsa,
+                                       padding);
+    if (ciphertxt_len <= 0)
+        goto err;
+    if (ciphertxt_len == plaintxt_len
+            && memcmp(decoded, plaintxt, plaintxt_len) == 0) {
+        goto err;
+    }
+
+    SELF_TEST_EVENT_oncorrupt_byte(&ev, ciphertxt);
+
+    decoded_len = RSA_private_decrypt(ciphertxt_len, ciphertxt, decoded, rsa,
+                                      padding);
+    if (decoded_len != plaintxt_len
+            || memcmp(decoded, plaintxt,  decoded_len) != 0) {
+        goto err;
+    }
+    ret = 0;
+err:
+    SELF_TEST_EVENT_onend(&ev, ret);
+    OPENSSL_free(ciphertxt);
+    return ret;
+}
+
+int SELF_TEST_keygen_pairwise_test_ecdsa(EC_KEY *eckey)
+{
+    int ret = 1;
+    unsigned char dgst[16] = {0};
+    unsigned int dgst_len = (unsigned int)sizeof(dgst);
+    ECDSA_SIG *sig = NULL;
+    ST_EVENT ev;
+
+    SELF_TEST_EVENT_init(&ev);
+    SELF_TEST_EVENT_onbegin(&ev, SELF_TEST_TYPE_PCT, SELF_TEST_DESC_PCT_ECDSA);
+
+    sig = ECDSA_do_sign(dgst, dgst_len, eckey);
+    if (sig == NULL)
+        goto err;
+
+    SELF_TEST_EVENT_oncorrupt_byte(&ev, dgst);
+
+    if (ECDSA_do_verify(dgst, dgst_len, sig, eckey) != 1)
+        goto err;
+
+    ret = 0;
+err:
+    SELF_TEST_EVENT_onend(&ev, ret);
+    ECDSA_SIG_free(sig);
+    return ret;
+}
+
+int SELF_TEST_keygen_pairwise_test_dsa(DSA *dsa)
+{
+    int ret = 1;
+    unsigned char dgst[16] = {0};
+    unsigned int dgst_len = (unsigned int)sizeof(dgst);
+    DSA_SIG *sig = NULL;
+    ST_EVENT ev;
+
+    SELF_TEST_EVENT_init(&ev);
+    SELF_TEST_EVENT_onbegin(&ev, SELF_TEST_TYPE_PCT, SELF_TEST_DESC_PCT_DSA);
+
+    sig = DSA_do_sign(dgst, (int)dgst_len, dsa);
```

Comment:
There seems to be confusion as to the correct type for dgst_len here. We calculate it using sizeof, which gives us a size_t. Then we cast it to an unsigned int. Then when we actually use it we cast it to an int. The final cast is explicit here, and implicit in the DSA_do_verify call below.